### PR TITLE
Add DragonFlyBSD support for dmd

### PR DESCRIPTION
**Pull request by dkgroot**
Bootstrapped via v2.067.1: [See](https://github.com/dkgroot/dragonflybsd_dmd_port)

All tests / unittests were completed successfully both on Linux as well as DragonFlyBSD after the port. (This was prior to rebasing, where 37d37a6 is currently failing)

Note: This port only supports 64-bit, because DragonFlyBSD is currently x86_64 only (i386 support removed a while back). This is also the reason why the dmd option '-m32'  is being suppressed when the target is dragonflybsd.

Links:
- [DragonFlyBSD Support](https://forum.dlang.org/post/CAG2UyHqsa7tv5cVMZWOrZXo5QQus66X5fmbBixxJLooH2J=eWw@mail.gmail.com)
- [Create a dmd-cxx branch on dlang/dmd](https://forum.dlang.org/post/CABOHX+ckJbL1yGyZBz-Lqq1kVWX=XDy50CT_H9=BNvuquP_Mmg@mail.gmail.com)

### CODE REVIEW

**Comment by dkgroot** [comment]
**Question Regarding Bootstrapping on DragonFlyBSD**
I settled on (after quite a bit of experimenting and back and forth [See](https://forum.dlang.org/post/CABOHX+ckJbL1yGyZBz-Lqq1kVWX=XDy50CT_H9=BNvuquP_Mmg@mail.gmail.com)): 
 - dmd v2.067.1
 - druntime dmd-cxx (port based on v2.067.1 also available)
 - phobos dmd-cxx (port based on v2.067.1 also available)

But i am not sure any of these branches are 'still' open for Pull Requests / Updates ?
If not, I could just provide the zip file with the dmd compiler, if requested.

Notes:
 - I could not get dmd-cxx to compile cleanly on linux (because of -fPIC / hardening issues on ubuntu). Fixing this takes quite a large number of changes and i did not want to mix PR's
 - I could not get dmd-cxx to compile the current master (2.077.1) on dragonflybsd (Lots of linking errors)
 - v2.067.1 worked flawlessly out of the box. 

Can one of you guide me what would be preferred ? Should we fix dmd-cxx first and then rebase my changes on top of it / just go with v2.067.1 for dragonflybsd changes.

File: test/d_do_test.d
Comment (wilzbach) [inline test/d_do_test.d]: Nit: Please add a comment that DragonFlyBSD is only `x86_64` here.

**Review by wilzbach — COMMENTED** [review]
Wow. First of all. Thanks a lot for your work. I start with first, quick review & tiny nit and answering your questions.

**Comment by wilzbach** [comment]
> But i am not sure any of these branches are 'still' open for Pull Requests / Updates ?

Of course. As explained on the mailing list, `dmd-cxx` has been created a couple of months ago & @ibuclaw has been working on backporting fixes to it since, e.g. his last PR: https://github.com/dlang/dmd/pull/7227

> Can one of you guide me what would be preferred ? Should we fix dmd-cxx first and then rebase my changes on top of it / just go with v2.067.1 for dragonflybsd changes.

tl:dr: how you bootstrap `DMD` is imho completely your own choice. As far as I can judge this doesn't affect "upstream" much, except you can / want to setup up a CI for DragonFlyBSD (which would be really cool). I would just go with the easiest way which seems to be v2.067.1 for now.
I don't know how package management works on DragonFlyBSD - will users be able to install DMD as binary package directly?

FYI: @WalterBright is pushing to convert the backend to D too (see e.g. https://github.com/dlang/dmd/pull/6907) and for this a couple of new features like `-betterC` would be required, which means there might be another bootstrapping waypoint.

The long-term plan was to maintain `dmd-cxx`, s.t. the life of porters is simplified and the number of bootstrapping points is reduced.
However, maintaining the backend and glue code is a lot of work and as no one has volunteered for it, so I doubt that that much work will happen here in the near future. `dmd-cxx` is based on `2.068.2`, contains all commits until the latest `cdmd` tag + Ian's backfixes. It would be cool to be able to compile `master` directly from `dmd-cxx`, but to be honest with the fast compilation time of DMD, it's a lot easier just use one more bootstrapping point and save all the hard energy required to add fixes and backports to `dmd-cxx` and invest it instead in building more cool things in D.
Also nowadays LDC is almost in sync with the latest DMD release and I presume it's a lot easier to use the LLVM's backend which supports cross-compilation for porting than bootstrapping through a couple of releases.

**Comment by dkgroot** [comment]
> > But i am not sure any of these branches are 'still' open for Pull Requests / Updates ?

> Of course. As explained on the mailing list, dmd-cxx has been created a couple of months ago & @ibuclaw has been working on backporting fixes to it since, e.g. his last PR: #7227

I read the mailinglist and forum post (Referenced in the PR).

> tl:dr: how you bootstrap DMD is imho completely your own choice. As far as I can judge this doesn't affect "upstream" much, except you can / want to setup up a CI for DragonFlyBSD (which would be really cool). 

I tried doing this on Travis ([See](https://travis-ci.org/dkgroot/dmd_dragonflybsd/builds/314543739), which worked, but was to slow, so it timed out during unittesting (mostly because running qemu-system on top of travis is really slow and i had to create a dragonflybsd installation (from iso) during testing). The process could have been split up in several stages (By storing the intermediate results somewhere). For now i gave up on this one.
I then tried using a x86_86-pc-dragonflybsd-gcc cross compiler (built for linux), but the dmd build system is not expecting a cross compiler (i think).
If you guys have a hosted environment (Like the auto tester uses), i would be happy to help setting it up.

> I would just go with the easiest way which seems to be v2.067.1 for now.

Ok once we conclude this PR for master I can push the PR for dragonflybsd support on v2.067.1. Seperating them might be a good idea, to keep focus. If you like me to push the PR right away, just let me know.

> I don't know how package management works on DragonFlyBSD - will users be able to install DMD as binary package directly?

DragonFlyBSD uses the 'Ports' from FreeBSD and does provide binary packages as well. Once we have all the PR's committed,I will update the FreeBSD ports Makefile to use the latest git revision. When FreeBSD accepts the updates, the results will automatically get picked up on DragonFlyBSD. It's a bit of a stack to work through, but we will get there.

> FYI: @WalterBright is pushing to convert the backend to D too (see e.g. #6907) and for this a couple of new features like -betterC would be required, which means there might be another bootstrapping waypoint.

No problem. Now that everything is working/running I already know where attention is needed.

File: test/d_do_test.d
Comment (jacob-carlborg) [inline test/d_do_test.d]: Curly braces on their own lines.

**Comment by wilzbach** [comment]
> I tried doing this on Travis (See, which worked, but was to slow, so it timed out during unittesting (mostly because running qemu-system on top of travis is really slow 

Hmm you could give CircleCi (2 vCPUs, 4GB) or Semaphore a try. They both are _a lot_ faster.
On CircleCi you can roll your own docker containers, maybe that already helps?
Alternatively, GDC for example, uses buildbot.

> If you guys have a hosted environment (Like the auto tester uses), i would be happy to help setting it up.

CC @braddr (the person in charge of auto-tester). 

It might also be possible to add more machines to the Jenkins CI (CC @MartinNowak).

> might be a good idea to have a wiki page about the porting process

Agreed and based on your recent experience you are an excellent candidate to start such a page ;-)

> Add bootstrap stage using dmd-cxx to all CI's, so that everyone is forced to keep it up date.

Currently the following CIs test bootstraping:
- auto-tester (uses v2.068.2 - last cdmd release)
- Travis (tests bootstrapping with gdc, ldc and dmd)

CircleCi, DAutotTest and Jenkins use a more recent, but fixed, DMD version.

Keeping `dmd-cxx` up to date is a good proposal, but as I said before, it needs an volunteer(s).

> backport the massive number of -fPIC PR's to dmd-cxx

I can help with that (it's just adding `-fPIC` flags to a lot of places), but FWIW we still can't  run parts of the test suite on a PIE hardened OS (though that's the last missing bit, see e.g. https://github.com/dlang/dmd/pull/7420)

**Comment by dkgroot** [comment]
> CC @braddr (the person in charge of auto-tester).
> It might also be possible to add more machines to the Jenkins CI (CC @MartinNowak).

I think, we should move this to a seperate PR when we get working on this. Thanks for the links though.

> Add bootstrap stage using dmd-cxx to all CI's, so that everyone is forced to keep it up date.

I was not expecting anything else :-)

> Currently the following CIs test bootstraping:
> - auto-tester (uses v2.068.2 - last cdmd release)
> - Travis (tests bootstrapping with gdc, ldc and dmd)
Auto Tester should not be a problem
Travis will be a bit of an issue, because of the speed. Running unittests in qemu is quite slow. Building bootstrap + master + druntime + phobos + all the unittests, is certainly going to time out.

> CircleCi, DAutotTest and Jenkins use a more recent, but fixed, DMD version.

Keeping dmd-cxx up to date is a good proposal, but as I said before, it needs an volunteer(s).

> I can help with that (it's just adding -fPIC flags to a lot of places), but FWIW we still can't run parts of the test suite on a PIE hardened OS (though that's the last missing bit, see e.g. #7420)

Once the dust settles on this (has only been a couple days) I will give it another shot.

**Comment by joakim-noah** [comment]
Nice work.  As for bootstrapping, the problem is that there's currently no strategy on how dmd wants to handle that for new platforms.  ldc, on the other hand, [puts out regular releases of its C++ ltsmaster branch, based on the dmd 2.068.2 frontend](https://github.com/ldc-developers/ldc/releases/tag/v0.17.5), which is continually tested on CI, so it's easy to do there.

If I were you, I'd focus on getting ldc ported first, then you could use your ldc package to build the latest dmd easily.  There is little value in porting the last C++ dmd, especially given the confusion about how dmd currently wants to deal with such ports.

Ideally, I'd like to see the dmd-cxx frontend updated as far as Iain got it, and then we could work to make sure it's a viable branch on all existing platforms, that could be used to bootstrap new x64 platforms.  But ldc ltsmaster is already here and it works.

**Comment by dkgroot** [comment]
@joakim-noah 
I full intend to get ldc working on DragonFly as well. Dave MacFarlane previously tried committing his DragonFly changes to ldc, but ran into problems (linking and needing upstream druntime/phobos changes committed).
I already have the dmd bootstrap working and all unittests are able to pass, so no real issue there. So the path taken is a little different but the end result will be the same.

**Comment by joakim-noah** [comment]
I understand that you already have it working, I'm only talking about with respect to getting stuff merged in the official D git repos and adding D packages to DragonFly ports, which you presumably would like to create without a bunch of local patches.

There is currently nowhere to merge changes to a C++ dmd, but there is for ldc, the ltsmaster branch.  In terms of getting these pulls merged into the master branch, there's no problem.

**Comment by dkgroot** [comment]
@joakim-noah 
Ok that sound like a good solution. I think porting the dragonfly changes to ldc:ltsmaster should be pretty straight forward. I already have the druntime/phobos:dmd-cxx branches ready, which i guess can be used directly (correct ?)

**Comment by joakim-noah** [comment]
ldc has forked [the 2.068.2 branches of druntime/phobos as ldc-ltsmaster](https://github.com/ldc-developers/druntime/tree/ldc-ltsmaster).  You may have a few merge conflicts, particularly if you are merging from a later release tag, but likely not many.

**Comment by dkgroot** [comment]
@joakim-noah
Regarding initial trial port on ldc / ltsmaster. Everything compiled cleanly. 
```
99% tests passed, 8 tests failed out of 663
Total Test time (real) =  59.37 sec
The following tests FAILED:
	227 - std.math (Failed)
	232 - std.numeric (Failed)
	304 - std.internal.math.gammafunction (Failed)
	552 - std.math-debug (Failed)
	557 - std.numeric-debug (Failed)
	629 - std.internal.math.gammafunction-debug (Failed)
	660 - dmd-testsuite-debug (Failed)
	662 - dmd-testsuite (Failed)
```
Not to bad i would say. There is something something funky with std/math.d, causing multiple unittests to fail, but have not been able to figure out what is causing this one yet :-)

**Comment by joakim-noah** [comment]
Look to all be math-related, my guess is that [DragonFly has the same `long double` issues as NetBSD](http://forum.dlang.org/thread/bqptgbdtekmqzbgenimo@forum.dlang.org), ie is `long double` 64-bit and DragonFly doesn't provide 80-bit math functions?

**Comment by dkgroot** [comment]
@joakim-noah 
Thanks for the links / reading material. Apparently this was not the issue. The mapping in core/stdc/math.d i used was wrong. Fixing those mappings yields:
```
99% tests passed, 4 tests failed out of 663
Total Test time (real) =  45.73 sec
The following tests FAILED:
	304 - std.internal.math.gammafunction (Failed)
	629 - std.internal.math.gammafunction-debug (Failed)
	660 - dmd-testsuite-debug (Failed)
	662 - dmd-testsuite (Failed)
```
Will have to copy these math changes to the druntime PR :-). Funny thing is the same implementation did not trigger any of these unittests to fail
.
Figuring out what is might be going wrong in std.internal.math.gammafunction is a little daunting. 
```
/root/dragonflybsd_dmd_port/ldc/ldc/runtime/phobos/std/internal/math/gammafunction.d(540): Assertion failure.
line 540 being: assert(logGamma(-50.2) == log(fabs(gamma(-50.2))));
```
BTW: We are getting a little offtopic here, regarding LDC. Can we maybe move this LDC related topid somewhere else ?

**Comment by joakim-noah** [comment]
If you plan on submitting a pull, [you can start with a WIP patch against the ldc-ltsmaster branch of druntime here](https://github.com/ldc-developers/druntime/pulls).  There's also [a forum for LDC, if you just want to discuss further](http://forum.dlang.org/group/ldc).

**Comment by dkgroot** [comment]
Regarding CI: I created a simple/example CI implementation to show:
 - Setting up dragonfly in a qemu/kvm based vmm on top of semaphore CI
 - Building dmd bootstrap (v2.067.1 / dmd-cxx) on dragonfly vm
 - Building dmd master on dragonfly vm
 - Running dmd / druntime / phobos master unittests
https://semaphoreci.com/dkgroot/dmd_dragonfly_ci
(Note this is just a reference/test/example project)

File: src/dmd/backend/backconfig.c
Comment (joakim-noah) [inline src/dmd/backend/backconfig.c]: It is ridiculous that every UNIX has to be added to this list so many places.  There needs to be a `TARGET_POSIX` and maybe other defines, so that this common list can be maintained in one header.  Not saying you should do it, just saying it shouldn't be this much work to add a new UNIX like Dragonfly.

Comment (ibuclaw) [reply src/dmd/backend/backconfig.c]: Backend can do as they please, but at some point I hope to remove all traces of `TARGET_XXX` in the frontend except the target module, where the logic will now live. (`global.params.isXXX` will be removed and made private vars in target also).

Comment (dkgroot) [reply src/dmd/backend/backconfig.c]: I do agree, especially when it comes to an OS that clearly comes from a particular family (BSD in this case). Initially I wanted to add a TARGET_BSD and have all BSD's inherit from that, which would simplify a lot of these cases.

I did contact a couple of people and checked other related PR's before going down this road, and this issue has been raised multiple/several times before. @WalterBright (amongst others) was strongly against mixing/reusing code and wants a separate definition for each architecture and os. 

This does have mean quite a bit of code duplication and added maintenance. But on the other hand, it does have a positive effect, namely changes made for DragonFly have 0 change of messing stuff up for FreeBSD, OpenBSD, Linux, Darwin etc. The latter was the main reason to implement it this way (If i am not mistaken).

Keeping this positive effect in place, there are still possibilities for refactoring. For example these large 'version(xxxx)' trees could be split into a single file per architecture and OS (You would still have code duplication but an easier job to port to a new OS. Namely: one single backend file to edit).

If you do see reason to re-open this discussion, i think it should take place elsewhere, and not this PR.

Comment (joakim-noah) [reply src/dmd/backend/backconfig.c]: >WalterBright (amongst others) was strongly against mixing/reusing code and wants a separate definition for each architecture and os.

I think you misunderstood.  Walter wants each OS to be versioned off separately for most D code, but clearly this compiler code is already not being separated, which is why you're just adding to this list.  I believe he'd be fine with merging all this into one define or eventually one D version, as Iain plans to do exactly that.

Anyway, I'm not reopening this discussion, as I noted I'm not asking you to consolidate, just making a comment that this needs to be changed, which Iain has acknowledged as the plan.

Comment (dkgroot) [reply src/dmd/backend/backconfig.c]: I see what you mean now. Adding a couple of generic TARGET groups (POSIX/BSD/WIN/OSX) would be quite nice, and clear up some of these long repetitive ifdefs. Hopefully this refactorying will happen after this PR has already been merged :-)

File: src/dmd/cli.d
Comment (dkgroot) [inline src/dmd/cli.d]: This spelling fix is unrelated to this PR

File: src/dmd/cli.d
Comment (dkgroot) [inline src/dmd/cli.d]: Not sure if this was an intended use ?
Not sure if it merits an explanation / comment ?

File: src/dmd/cli.d
Comment (dkgroot) [inline src/dmd/cli.d]: These questions/comments are not really related to this PR specifically:
What about the other OS's like NetBSD / OpenBSD etc ?
Would adding posix make sense ? 
Do we really need yet another spelling / enum for the target OS ?

Comment (wilzbach) [reply src/dmd/cli.d]: Well, the problem is that `dlang.org` needs to be able to set the target OS and the current `TARGET_` are constants. I'm not really happy with this approach.
@ZombineDev wanted to do a more general revamp and introduce target triples. In any case, I think currently there are only differences between `Windows` and `Posix`, so it might make sense to just use those two.

File: src/dmd/cli.d
Comment (wilzbach) [reply src/dmd/cli.d]: Yes that's how it was intended to be used.
BTW `-m32` is also mostly defunct on ArchLinux too as they dropped native support for 32-bit too, but I'm not sure whether it makes sense to complicate the code with such subtleties.

File: src/dmd/cli.d
Comment (dkgroot) [reply src/dmd/cli.d]: @wilzbach : I did not see the original PR, or i would have made these comments there. 

The scattered nature and their repetition of these OS specific definitions is what throws me off a little to be honest. From a porting perspective it would be nice to see all OS specific stuff in one single file / directory, but that has been discussed before (over and over/ad nausea-um), so dropping that.

Except i just added one more 'exception' in line 215 (see above) so now there is three :-)

File: src/dmd/cli.d
Comment (dkgroot) [reply src/dmd/cli.d]: Using -m32 with dmd on dragonfly currently does fail (not implemented -> assertion failure). I guess this is not the same on archlinux. They might just lack the 32-bit libraries to compile against.
I can remove it if you want (Or wait for another reviewer).

Comment (wilzbach) [reply src/dmd/cli.d]: Yes, you can still compile `gcc-multilib` yourself from the AUR (arch user repository).
I don't this being used like you did as I said before, that was intended ;-)

Comment (dkgroot) [reply src/dmd/cli.d]: OK issue closed

**Comment by dkgroot** [comment]
Hi guys, would it help if this PR were be split by file, so that it would make a little more progress ?

**Comment by joakim-noah** [comment]
No, that would make things worse.  It just needs to be reviewed, I'll do it.

**Comment by dkgroot** [comment]
@joakim-noah Ok thanks for letting me know. I just noticed the speed with which the musl PR's went through. good that I asked.

File: src/dmd/backend/os.c
Comment (joakim-noah) [inline src/dmd/backend/os.c]: Regex failure?  Two more below.

File: src/dmd/mars.d
Comment (joakim-noah) [inline src/dmd/mars.d]: This doesn't exist, would rather you removed the nonexistent OpenBSD link above, as we don't put out official releases for these platforms.

**Review by joakim-noah — CHANGES_REQUESTED** [review]
Four small nits, otherwise fine.

File: src/dmd/backend/cdef.h
Comment (dkgroot) [reply src/dmd/backend/cdef.h]: Fixed
Also fixed a faulty reference to "TARGET_OPENBSD" in the OpenBSD entry above (in a seperate commit).

**Review by joakim-noah — APPROVED** [review]
Looks good.

**Review by PetarKirov — APPROVED** [review]

**Comment by ZombineDev** [comment]
Nice work, @dkgroot! Congratulations for getting through all the necessary steps!

File: ini/dragonflybsd/bin64/dmd.conf
Comment (dkgroot) [reply ini/dragonflybsd/bin64/dmd.conf]: @ZombineDev I will add -fPIC, even though it is only actually required on (most) Linux versions. FYI: checked the other platforms' ini-files and they also do not have -fPIC added by default.

**Comment by dkgroot** [comment]
Thanks everyone for your reviews, hints and time !

**Pull request merged**
